### PR TITLE
Security

### PR DESCRIPTION
**Pull request by bigbrownking**
Configured CORS policy
Added Brute Force Protector
Added BCryptPasswordEncoder

### CODE REVIEW

File: src/main/java/org/example/config/SecurityConfig.java [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+                          JwtRequestFilter jwtRequestFilter) {
+        this.userDetailsService = userDetailsService;
+        this.jwtRequestFilter = jwtRequestFilter;
+    }
+
+    @Bean
+    public SecurityFilterChain securityFilterChain(HttpSecurity http) throws Exception {
+        http
+                .csrf().disable()
+                .cors()
```
Comment (sergeypil) [inline src/main/java/org/example/config/SecurityConfig.java]: These two methods are depricated. Use
```suggestion
                .cors(cors -> cors.configurationSource(corsConfigurationSource()))
                .csrf(AbstractHttpConfigurer::disable)
```

File: src/main/java/org/example/config/SecurityConfig.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+                        .requestMatchers("/trainee/**").hasRole("TRAINEE")
+                        .requestMatchers("/trainer/**").hasRole("TRAINER")
+                        .anyRequest().authenticated()
+                )
+                .userDetailsService(userDetailsService)
+                .sessionManagement(session -> session
+                        .sessionCreationPolicy(SessionCreationPolicy.STATELESS)
+                )
+                .addFilterBefore(jwtRequestFilter, UsernamePasswordAuthenticationFilter.class)
+                .formLogin(formLogin -> formLogin
```
Comment (sergeypil) [inline src/main/java/org/example/config/SecurityConfig.java]: Why did you add formLogin? Does it work together with AuthController.login?

File: src/main/java/org/example/controller/AuthController.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    @PostMapping("/login")
+    public ResponseEntity<String> login(@RequestBody LoginRequestDto loginRequestDto) {
+        String username = loginRequestDto.getUsername();
+
+        if (loginAttemptService.isBlocked(username)) {
+            return new ResponseEntity<>("User is blocked due to too many login attempts. Please try again later.", HttpStatus.LOCKED);
+        }
+
+        try {
+            Authentication authentication = authenticationManager.authenticate(
```
Comment (sergeypil) [inline src/main/java/org/example/controller/AuthController.java:61]: Seems like we need to authentication object to securityContextHolder

File: src/main/java/org/example/controller/TraineeController.java [hunk 4]
```diff
@@ -160,13 +166,16 @@ public ResponseEntity<Void> activateTrainee(@RequestBody ActivateUserRequestDto
...
     @ApiResponses(value = {
             @ApiResponse(responseCode = "200", description = "Trainee deactivated successfully"),
             @ApiResponse(responseCode = "500", description = "Internal server error")
     })
     @PatchMapping("/deactivate")
-    public ResponseEntity<Void> deactivateTrainee(@RequestBody DeactivateUserRequestDto deactivateUserRequestDto) {
+    public ResponseEntity<Void> deactivateTrainee(
+            @AuthenticationPrincipal UserDetails userDetails,
+            @RequestBody DeactivateUserRequestDto deactivateUserRequestDto) {
         try {
```
Comment (sergeypil) [inline src/main/java/org/example/controller/TraineeController.java]: We should not catch and process exceptions like this in controller. We should exception handler. For example using @RestControllerAdvice and @ExceptionHandler

File: src/main/java/org/example/controller/TraineeController.java [hunk 5]
```diff
@@ -66,7 +55,9 @@ public ResponseEntity<Void> login(@RequestBody LoginRequestDto loginRequestDto)
             @ApiResponse(responseCode = "500", description = "Internal server error")
     })
     @PutMapping("/update")
-    public ResponseEntity<UpdateTraineeResponseDto> updateTrainee(@RequestBody UpdateTraineeRequestDto updateTraineeRequestDto) {
+    public ResponseEntity<UpdateTraineeResponseDto> updateTrainee(
+            @AuthenticationPrincipal UserDetails userDetails,
```
Comment (sergeypil) [inline src/main/java/org/example/controller/TraineeController.java]: Why do you have userDetails as parameter if you do not use it?

File: src/main/java/org/example/controller/TrainingController.java [hunk 6]
```diff
@@ -34,26 +36,20 @@ public TrainingController(TrainingService trainingService) {
...
-            return ResponseEntity.ok(response);
-        } catch (Exception e) {
-            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).build();
-        }
+        GetTrainingTypesResponseDto response = trainingService.getTrainingTypes();
+        return ResponseEntity.ok(response);
     }
 
+
+    @PreAuthorize("isAuthenticated()")
```
Comment (sergeypil) [inline src/main/java/org/example/controller/TrainingController.java]: Do we really need @PreAuthorize("isAuthenticated()") for register method?

**Review by sergeypil — APPROVED** [review]

File: src/main/java/org/example/controller/TrainingController.java (hunk 6 above)
Comment (bigbrownking) [reply src/main/java/org/example/controller/TrainingController.java]: Oops🤭
